### ISSUE TITLE
Lambda events bootstrap

### ISSUE THREAD

**Issue by IndikaUdagedara** [aws/aws-lambda-dotnet#556]
Bootstrapping an APIGateway request handling lambda and its interoperability with the ASP.NET MVC pipeline is a breeze due to the [APIGatewayProxyFunction ](https://github.com/aws/aws-lambda-dotnet/blob/master/Libraries/src/Amazon.Lambda.AspNetCoreServer/APIGatewayProxyFunction.cs).

I found it a good pattern to follow with other integrations as well (e.g. SQS, APICustomAuthorizer etc.) since it

- gives the ability to use ASP.NET utilities (configuration, DI, logging) seamlessly
- enables local exploratory testing e.g. hit an API endpoint and pass integration requests

I got this working without much trouble again thanks to the `LambdaServer` and other extension points in `Amazon.Lambda.AspNetCoreServer`.  Do you think it would be useful to be added in this repo? I made a start at [https://github.com/IndikaUdagedara/dotnet-lambda-extensions](https://github.com/IndikaUdagedara/dotnet-lambda-extensions). I can create a PR if you think it's a good idea.

**Comment by normj** [comment]
I'm intrigued by the idea. I would love to hear more thoughts from others in the community. Tagging some others that have commented on issues/PR for Amazon.Lambda.AspNetCoreServer in the past to solicited feedback.

@arndklocker @andymac4182 @martonx @psbmiller @aidansteele @chrisoverzero

**Comment by chrisoverzero** [comment]
I couldn’t say that I disagree, given that I maintain [a library][] for my company to do nearly this same thing. All of ASP.NET Core shouldn’t be _necessary_ for it, I think? We’ve found that `sam local invoke` combined with the .NET Core generic host gets us there. (This gets even nicer with 3.1, I seem to recall, as the host ecosystem for web and not-web are unified.) But maybe I’m missing something by only skimming; I can read this repo in detail during the week. 

I am excited at the possibility a more .NET-idiomatic Lambda experience, irrespective of the exact form it takes, to be sure!

[a library]: https://github.com/Cimpress-MCP/Tiger-Lambda

**Comment by IndikaUdagedara** [comment]
Yes, all of  MVC (`.AddMvc()`) is not necessary - only the following should be sufficient

```
                .AddMvcCore()
                .AddApiExplorer()
                .AddJsonFormatters();
```

Generic host would be better (and lightweight, I _assume_) but as far as I understood, most of `Amazon.Lambda.AspNetCoreServer` ([AbstractAspNetCoreFunction.cs](https://github.com/aws/aws-lambda-dotnet/blob/master/Libraries/src/Amazon.Lambda.AspNetCoreServer/AbstractAspNetCoreFunction.cs)) is coupled with WebHost so just wondering if it would be possible without a major refactor (not sure if you already have this planned in preparation for 3.x unified host)

**Comment by IndikaUdagedara** [comment]
Hi @normj, wanted to have another go at this after noticing the library now supports `GenericHost`.

This is what I'd propose:

1. A .NET idiomatic experience by providing a `Lambda` integration to the application host e.g. `IHostBuilder.UseLambda(...)`
2. A type-qualified function handler that provides an abstraction over the _low-level_ handler (similar to what `APIGatewayProxyFunction` provides)
e.g. `TResponse ILambdaFunction<TRequest, TResponse>:Handle(TRequest)`
3. To simplify local invocation, expose a web endpoint so Lambda handler can be invoked locally with common tools. This would be something like `IHostBuilder.UseLambdaLocal()` . To support `sam local` something like `IHostBuilder.UseLambdaSamLocal()` can be added.

I did a POC [here](https://github.com/IndikaUdagedara/dotnet-lambda-extensions) which is basically few extension methods on `HostBuilder` and an abstract class similar to `AbstractAspNetCoreFunction` but probably bit more abstract (😉) which I called `AbstractDotNetCoreFunction`... `AbstractAspNetCoreFunction` could extend that and implement the WebHost specifics using `LambdaServer` etc. Other non-web-like event handlers could just implement `AbstractDotNetCoreFunction` (my sample has a `LambdaAuthorizer`). 

What are your thoughts? I'm happy to do a PR if you think the proposed implementation is sound..

Thanks.
